### PR TITLE
[MplCore] Deposit sell

### PR DESCRIPTION
**Pull request by JeremyLi28**
Added new `mpl_core_deposit_sell.rs` instruction and sdk + tests

* Added a new allowlist type `update_authority` to be used by mpl core collections. 
* Mpl core assets without collection can still use mintAddress allowlist type

### CODE REVIEW

File: programs/mmm/src/util.rs [hunk 1]
```diff
@@ -783,6 +785,68 @@ pub fn check_allowlists_for_mint_ext(
     Err(MMMErrorCode::InvalidAllowLists.into())
 }
 
+pub fn check_allowlists_for_mpl_core<'info>(
+    allowlists: &[Allowlist],
+    asset: &Box<Account<'info, IndexableAsset>>,
```
Comment (madergaser) [inline programs/mmm/src/util.rs:790]: nit: should be able to just 
```suggestion
    asset: &IndexableAsset,
```

File: programs/mmm/src/instructions/mpl_core_asset/mpl_core_deposit_sell.rs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    )]
+    pub sell_state: Account<'info, SellState>,
+    /// CHECK: check collection later
+    collection: UncheckedAccount<'info>,
+
+    pub system_program: Program<'info, System>,
+    pub asset_program: Interface<'info, AssetInterface>,
+}
+
+pub fn handler(ctx: Context<MplCoreDepositSell>, args: DepositSellArgs) -> Result<()> {
```
Comment (nothing0012) [inline programs/mmm/src/instructions/mpl_core_asset/mpl_core_deposit_sell.rs:52]: let's define a new args type for this, and we don't need amount

File: programs/mmm/src/state.rs [hunk 3]
```diff
@@ -11,6 +11,8 @@ pub const ALLOWLIST_KIND_MINT: u8 = 2;
 pub const ALLOWLIST_KIND_MCC: u8 = 3;
 pub const ALLOWLIST_KIND_METADATA: u8 = 4;
 pub const ALLOWLIST_KIND_GROUP: u8 = 5;
+// this is should only be used by mpl core with update_authority as collection.
+pub const ALLOWLIST_KIND_UPDATE_AUTHORITY: u8 = 6;
```
Comment (nothing0012) [inline programs/mmm/src/state.rs:15]: rename this to be `ALLOWLIST_KIND_MPL_CORE_COLLECTION`

File: programs/mmm/src/util.rs [hunk 4]
```diff
@@ -783,6 +785,68 @@ pub fn check_allowlists_for_mint_ext(
...
+    allowlist_aux: Option<String>,
+) -> Result<()> {
+    if allowlists
+        .iter()
+        .any(|&val| val.kind == ALLOWLIST_KIND_METADATA)
```
Comment (nothing0012) [inline programs/mmm/src/util.rs:795]: explore if we can make Core's onchain traits verifiable.

File: programs/mmm/src/util.rs [hunk 5]
```diff
@@ -783,6 +785,68 @@ pub fn check_allowlists_for_mint_ext(
...
+                        return Err(MMMErrorCode::InvalidAllowLists.into());
+                    }
+                    return Ok(());
+                } else {
+                    return Err(MMMErrorCode::InvalidAllowLists.into());
+                }
+            }
+            ALLOWLIST_KIND_MINT => {
+                if asset.key() != allowlist_val.value {
+                    return Err(MMMErrorCode::InvalidAllowLists.into());
```
Comment (nothing0012) [inline programs/mmm/src/util.rs:832]: maybe fold this into `_`, because this only applies to ONE asset.

File: sdk/src/constants.ts [hunk 6]
```diff
@@ -18,6 +18,7 @@ export enum AllowlistKind {
   mcc = 3,
   metadata = 4,
   group = 5,
+  update_authority = 6,
```
Comment (nothing0012) [inline sdk/src/constants.ts:21]: also change the enum here

File: sdk/src/mmmClient.ts [hunk 7]
```diff
@@ -58,6 +64,26 @@ export const MMMProgramID = new PublicKey(
   'mmm3XBJg5gk8XJxEKBvdgptZz6SgK4tXvn36sodowMc',
 );
 
+export function isMplCoreAsset(
+  account: anchor.web3.AccountInfo<Buffer>,
+): boolean {
+  return account?.owner.equals(new PublicKey(MPL_CORE_PROGRAM_ID)) ?? false;
```
Comment (nothing0012) [inline sdk/src/mmmClient.ts:70]: MPL_CORE_PROGRAM_ID is already a PublicKey, right? so we don't need a `new` here

File: tests/mmm-mpl-core.spec.ts [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+      assert.strictEqual(err.code, 6001);
+    }
+
+    const refreshedAsset = await getTestMplCoreAsset(asset.publicKey);
+
+    // Verify asset account.
+    assert.equal(refreshedAsset.owner.toString(), wallet.publicKey.toString());
+  });
+
+  it("can't deposit sell - no collection", async () => {
```
Comment (nothing0012) [inline tests/mmm-mpl-core.spec.ts:171]: name fix: allowlist mismatch

File: sdk/src/mmmClient.ts (hunk 7 above)
Comment (JeremyLi28) [reply sdk/src/mmmClient.ts:70]: it is a separate PublicKey type defined in mpl core package, need to convert to web3 type here

File: programs/mmm/src/util.rs (hunk 1 above)
Comment (JeremyLi28) [reply programs/mmm/src/util.rs:790]: interesting,

File: programs/mmm/src/instructions/mpl_core_asset/mpl_core_deposit_sell.rs [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+        mut,
+        seeds = [POOL_PREFIX.as_bytes(), owner.key().as_ref(), pool.uuid.as_ref()],
+        has_one = owner @ MMMErrorCode::InvalidOwner,
+        has_one = cosigner @ MMMErrorCode::InvalidCosigner,
+        bump
+    )]
+    pub pool: Box<Account<'info, Pool>>,
+    #[account(
+        mut,
+        constraint = asset.to_account_info().owner == asset_program.key,
```
[the remark this answers is not in the document]
Comment (JeremyLi28) [reply programs/mmm/src/instructions/mpl_core_asset/mpl_core_deposit_sell.rs:35]: will reply on cpi check for now

File: programs/mmm/src/util.rs (hunk 4 above)
Comment (JeremyLi28) [reply programs/mmm/src/util.rs:795]: will be address in the future when we support onchain attribute offer for mpl core

File: programs/mmm/src/instructions/mpl_core_asset/mpl_core_deposit_sell.rs [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+    let asset = &ctx.accounts.asset;
+    let pool = &mut ctx.accounts.pool;
+    let sell_state = &mut ctx.accounts.sell_state;
+    let collection = &ctx.accounts.collection;
+
+    if pool.using_shared_escrow() {
+        return Err(MMMErrorCode::InvalidAccountState.into());
+    }
+
+    let _ = check_allowlists_for_mpl_core(&pool.allowlists, asset, args.allowlist_aux)?;
```
Comment (solonk8) [inline programs/mmm/src/instructions/mpl_core_asset/mpl_core_deposit_sell.rs:68]: nit: don't need `let _ =`

**Review by solonk8 — APPROVED** [review]

File: programs/mmm/src/instructions/mpl_core_asset/mpl_core_deposit_sell.rs (hunk 10 above)
Comment (JeremyLi28) [reply programs/mmm/src/instructions/mpl_core_asset/mpl_core_deposit_sell.rs:68]: compiler seems will show some warning if I remove them

**Pull request merged**